src/profilers.cpp

```diff
@@ -2,19 +2,373 @@
+
+    fs::path stderr_metric_command;
+    stderr_metric_command = result_out / "metric_command_stderr.log";
+
+    const int ERROR_STDERR = 201;
+    const int ERROR_STDOUT_DUP2 = 202;
+    const int ERROR_STDERR_DUP2 = 203;
+    const int ERROR_NO_NUMBER_REGEX = 204;
+    const int ERROR_TOO_MANY_NUMBERS_REGEX = 205;
+    const int ERROR_CONVERSION_TO_FLOAT = 206;
+    const int ERROR_PARSING_CONNECTION_INSTRS = 207;
+    const int ERROR_USER_REGEX_MATCH = 208;
+    
+    
+    std::vector<std::string> parts = boost::program_options::split_unix(this->metric_command);
+
+    const char *path = parts[0].c_str();
+    char *parts_c_str[parts.size() + 1];
+
+    for (int i = 0; i < parts.size(); i++) {
+      parts_c_str[i] = (char *)parts[i].c_str();
+    }
+    parts_c_str[parts.size()] = NULL;
+
+
+    pid_t forked_metric_profiler = fork(); //separate process for profiler, parent handles errors
+
+    if(forked_metric_profiler == 0){
+      int code_metric_exec = 0;
+
+      int status;
+      pid_t command_status = waitpid(pid, &status, WNOHANG); //check if profiled command has finished executing
+      const int interval_ns = 1000 * 1000000 / this->freq;
+
+      while (command_status > 0){
+
+        struct timespec start, end;
+        clock_gettime(CLOCK_MONOTONIC, &start);
+
+        int pipe_fd[2];
+
+        if (pipe(pipe_fd) == -1) {
+          print("Could not establish communication pipe between metric-exec "
+                "and metric-read in \"" + this->get_name() + "\"! Terminating "
+                "the profiled command wrapper.", true, true);
+          kill(pid, SIGTERM);
+          return;
+        }
+
+        pid_t forked_metric_exec = fork();
+        if (forked_metric_exec == 0){
+          close(pipe_fd[0]);
+
+          fs::current_path(result_processed);
+
+          int stderr_fd = creat(stderr_metric_command.c_str(), O_WRONLY);
+
+          if (stderr_fd == -1) {
+            std::exit(ERROR_STDERR);
+          }
+
+          if (dup2(stderr_fd, STDERR_FILENO) == -1) {
+            std::exit(ERROR_STDERR_DUP2);
+          }
+
+          if (dup2(pipe_fd[1], STDOUT_FILENO) == -1) {
+            std::exit(ERROR_STDOUT_DUP2);
+          }
+
+          close(pipe_fd[1]);
+          execvp(path, parts_c_str);
+
+          std::exit(errno);
+        }
+        else
+        {
+          close(pipe_fd[1]);
+          //read
+          char buffer[1024];
+          ssize_t bytes_read;
+          std::string data;
+          
+          while ((bytes_read = read(pipe_fd[0], buffer, sizeof(buffer) - 1)) > 0) {
+              buffer[bytes_read] = '\0';
+              data += buffer;
+          }
+          std::string parsed_data;
+
+          if(!this->regex.empty()){
+
+            std::regex regexPattern(this->regex);
+            std::smatch match_pattern;
+            if (std::regex_search(data, match_pattern, regexPattern)) {
+                parsed_data = match_pattern.str();
```

Comment:
You should assume that the correct regex has a single capturing group (e.g. "blabla(\d+)"). This line may not process it correctly as it will assign an entire match to ```parsed_data``` rather than just the contents of the capturing group as far as I know. Also, consider what will happen when a non-conforming regex is provided by the user (e.g. "blabla(\d+)(\d+)" with too many capturing groups).

Reply:
I don't know if what I did was the most efficient way to solve the problem. But I used the user input regex as a first filter, and on that first match I made another regex that counts how many floats I have. If there are 0 or more than one, I exit with specific error codes.